### PR TITLE
Implement test resources property provider factory

### PR DESCRIPTION
**Pull request by melix**
This commit introduces a new annotation, `@TestResourceProperties`, which can be applied on a test in order to tell that the test needs to resolve one or more test resource properties _before_ the application context is available.

For example, with:

```java
@MicronautTest
@TestResourcesProperties(
    value = "redis.url"
)
class SomeTest { ... }
```

Then the value of the `redis.url` property will automatically be fetched from the test resources service and made available as if it had been done in a `TestPropertyProvider`. The difference is that this without this annotation, it was required to call the test resources client directly in the provider, which was error prone. In particular, the `properties` argument of the `resolve` call are not easy to figure out.

This annotation is therefore a convention to avoid having to call the client directly. However, in some cases it might be necessary to read a property in order to compute a different one which needs to be available in the test.

This can be done by adding a `providers` argument to the annotation:

```java
@TestResourcesProperties(
    value = "rabbitmq.uri",
    providers = RabbitTest.RabbitMQProvider.class
)
class RabbitTest {
  // ...
      @ReflectiveAccess
      public static class RabbitMQProvider implements TestResourcesPropertyProvider {
          @Override
          public Map<String, String> provide(Map<String, Object> testProperties) {
              String uri = (String) testProperties.get("rabbitmq.uri");
              return Map.of(
                  "rabbitmq.servers.product-cluster.port", String.valueOf(URI.create(uri).getPort())
              );
          }
      }
}
```

The `TestResourcesPropertyProvider` type has access to all properties which are already resolved at the moment it is called, which includes the properties asked in `@TestResourcesProperties`.

This should make the implementation of https://github.com/micronaut-projects/micronaut-nats/pull/321 easier.

### CODE REVIEW

File: test-resources-extensions/test-resources-extensions-core/src/test/resources/application-test.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+test-resources:
```
Comment (graemerocher) [inline test-resources-extensions/test-resources-extensions-core/src/test/resources/application-test.yml:1]: is this used somewhere?

Comment (melix) [reply test-resources-extensions/test-resources-extensions-core/src/test/resources/application-test.yml:1]: Missed a test, added

File: test-resources-extensions/test-resources-extensions-core/src/main/java/io/micronaut/test/extensions/testresources/TestResourcesPropertyProvider.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.test.extensions.testresources;
+
+import java.util.Map;
+
+@FunctionalInterface
+public interface TestResourcesPropertyProvider {
```
[the remark this answers is not in the document]
Comment (melix) [reply test-resources-extensions/test-resources-extensions-core/src/main/java/io/micronaut/test/extensions/testresources/TestResourcesPropertyProvider.java:39]: added

File: test-resources-junit5/src/main/java/io/micronaut/test/extensions/testresources/annotation/TestProperty.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+package io.micronaut.test.extensions.testresources.annotation;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.ANNOTATION_TYPE, ElementType.FIELD})
+public @interface TestProperty {
```
[the remark this answers is not in the document]
Comment (melix) [reply test-resources-junit5/src/main/java/io/micronaut/test/extensions/testresources/annotation/TestProperty.java]: This wasn't actually supposed to be part of that PR :D

File: test-resources-extensions/test-resources-extensions-core/src/main/java/io/micronaut/test/extensions/testresources/annotation/TestResourcesProperties.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import io.micronaut.test.extensions.testresources.TestResourcesPropertyProvider;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.ANNOTATION_TYPE, ElementType.TYPE})
+public @interface TestResourcesProperties {
```
[the remark this answers is not in the document]
Comment (melix) [reply test-resources-extensions/test-resources-extensions-core/src/main/java/io/micronaut/test/extensions/testresources/annotation/TestResourcesProperties.java:37]: added

**Review by graemerocher — APPROVED** [review]

**Comment by graemerocher** [comment]
Some reference docs would be nice

**Comment by melix** [comment]
> Some reference docs would be nice

Yes, but I'd like to finish the improvements first (there should be another PR on this topic, with scope management).

**Pull request merged**